### PR TITLE
Support Microsoft Azure Storage Blobs

### PR DESCRIPTION
**Pull request by jpoon**
First PR out of many to add Azure support. Partially implements https://github.com/pachyderm/pachyderm/issues/960 adding ability to write to Microsoft Azure Storage Blobs for the object store.

* Add vendored dependency against Azure Storage SDK (github.com/Azure/azure-sdk-for-go/storage)
* Azure has a similar concept to buckets and they are called *containers*
* Not entirely sure how to add any automated tests around this as it would involve either running the emulator (which I don't think would be possible in Travis) or bringing up an actual cluster in Azure.

### CODE REVIEW

**Comment by jpoon** [comment]
Initially failed with:

```
--- FAIL: TestFlushCommitAfterCreatePipeline (6.74s)
	require.go:153: /home/travis/gopath/src/github.com/pachyderm/pachyderm/src/server/pachyderm_test.go:2006
	require.go:155: No error is expected but got commit TestFlushCommitAfterCreatePipelinePipelinef2cf16a68ae4/TestFlushCommitAfterCreatePipelinePipelinef2cf16a68ae4:79b21857023448f58a058fab7db0a61d:9 was cancelled
```

I can't see how my changes could have affected a hyper-kube deployment. Pushing a random change to force another build. *fingers crossed* hoping the earlier failure was a transient one.

File: src/server/pkg/deploy/assets/assets.go [hunk 1]
```diff
@@ -411,7 +427,8 @@ func RethinkRc(backend backend, volume string, hostPath string) *api.Replication
 		spec.Spec.Template.Spec.Volumes[0].PersistentVolumeClaim = &api.PersistentVolumeClaimVolumeSource{
 			ClaimName: rethinkVolumeClaimName,
 		}
-	} else if backend == localBackend {
+	} else if backend == localBackend || backend == microsoftBackend {
+		// ToDo: workaround until https://github.com/pachyderm/pachyderm/issues/960
```
Comment (jdoliner) [inline src/server/pkg/deploy/assets/assets.go]: Little confused by this comment, #960 is the Azure Support issue so it seems like this PR is kinda dependent on itself. Wasn't there something about needing k8s 1.4 to do provide data disks to rethink/etcd? Is that what you're referencing here?

**Comment by jdoliner** [comment]
@jpoon awesome! In terms of automated tests, our existing set of tests should be able to provide pretty good coverage on the code. Of course the difficulty is deploying Pachyderm with access to Azure Storage in CI. We haven't quite figured that out for s3 and GCS either so for the time being it's something we'll just have to live with. That error you saw is one I see from time to time so I believe it to be transient.
Looks like this is pretty much ready for review and then merge or is there more still to be done?

**Comment by jpoon** [comment]
Nah, that it's for this PR. I'm planning to split the work for #960 into several PRs: 

* this one (#969) adds support for writing/reading to Azure blob storage
* separate PR for using an Azure data disk for rethinkdb/etcd
* documentation changes/make targets, etc

Let me know if you want me to make a giant one instead.

File: src/server/pkg/deploy/assets/assets.go (hunk 1 above)
Comment (jpoon) [inline src/server/pkg/deploy/assets/assets.go]: Oops. Sorry for the confusion. Yep, workaround until k8s 1.4 + Azure data-disk support (which is the next PR I'll work on)

Actually, not sure if this change is entirely necessary but it's purpose was to use the local host path for the time being.

Comment (jdoliner) [inline src/server/pkg/deploy/assets/assets.go]: Yeah, that's a totally reasonable thing to do for the time being to get things testable.

**Comment by jdoliner** [comment]
Awesome, we generally prefer small PRs to big ones too so this works great for us. Will dive in and review now.

**Comment by jdoliner** [comment]
Alright, this all LGTM.
CLA is already signed so merging.
We like to do LGTM with memes, so please see below.
g1463594680341105745

**Pull request merged**